cybozu-go/moco
# Resizing PVC

refs: https://github.com/cybozu-go/moco/issues/265

Add enhancement to ｓupport applying PVC template changes to StatefulSet.
See https://github.com/cybozu-go/moco/pull/412 for design docs.

This PR implements the core functionality, but the following elements are done in a separate pull request


***

Manually tested before adding E2E test:

<details>

1. apply mysqlcluster

```console
$ cat mycluster.yaml
apiVersion: moco.cybozu.com/v1beta1
kind: MySQLCluster
metadata:
  namespace: default
  name: test
spec:
  replicas: 3
  podTemplate:
    spec:
      containers:
      - name: mysqld
        image: quay.io/cybozu/mysql:8.0.28
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes: [ "ReadWriteOnce" ]
      resources:
        requests:
          storage: 1Gi

$ kubectl apply -f mycluster.yaml
mysqlcluster.moco.cybozu.com/test created
```

2. check PVC

```conosole
$ kubectl get pvc -o yaml | grep -C 2 requests
    - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
    storageClassName: standard
--
    - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
    storageClassName: standard
--
    - ReadWriteOnce
    resources:
      requests:
        storage: 1Gi
    storageClassName: standard
```

3. check statefulset

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: moco-test
  namespace: default
spec:
<snip>
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      name: mysql-data
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
      volumeMode: Filesystem
```

4. edit mysqlcluster

```diff
  apiVersion: moco.cybozu.com/v1beta1
  kind: MySQLCluster
  metadata:
    namespace: default
    name: test
  spec:
    replicas: 3
    podTemplate:
      spec:
        containers:
        - name: mysqld
          image: quay.io/cybozu/mysql:8.0.28
    volumeClaimTemplates:
    - metadata:
        name: mysql-data
+       labels:
+         foo: bar
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
+           storage: 2Gi
-           storage: 1Gi
```

5. check PVC

```
$ kubectl get pvc -o yaml | grep -C 2 requests
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
    storageClassName: standard
--
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
    storageClassName: standard
--
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
    storageClassName: standard
```

6. check sts

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: moco-test
  namespace: default
spec:
<snip>
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      labels:
        foo: bar
      name: mysql-data
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 2Gi
      volumeMode: Filesystem
```


</details>
